corporate/views/remote_billing_page.py

```diff
@@ -0,0 +1,175 @@
+    remote_server_uuid = identity_dict["remote_server_uuid"]
+    remote_realm_uuid = identity_dict["remote_realm_uuid"]
+
+    try:
+        remote_server = RemoteZulipServer.objects.get(uuid=remote_server_uuid)
+    except RemoteZulipServer.DoesNotExist:
+        raise JsonableError(_("Invalid remote server."))
+
+    try:
+        remote_realm = RemoteRealm.objects.get(uuid=remote_realm_uuid, server=remote_server)
+    except RemoteRealm.DoesNotExist:
+        raise AssertionError(
+            "The remote realm is missing despite being in the RemoteBillingIdentityDict"
+        )
+
+    remote_server_and_realm_info = {
+        "remote_server_uuid": remote_server_uuid,
+        "remote_server_hostname": remote_server.hostname,
+        "remote_server_contact_email": remote_server.contact_email,
+        "remote_server_plan_type": remote_server.plan_type,
+        "remote_realm_uuid": remote_realm_uuid,
+        "remote_realm_host": remote_realm.host,
+    }
+
+    return render(
+        request,
+        "corporate/remote_billing.html",
+        context={
+            "user_email": user_email,
+            "user_full_name": user_full_name,
+            "remote_server_and_realm_info": remote_server_and_realm_info,
+        },
+    )
+
+
+def remote_billing_plans_common(
+    request: HttpRequest, realm_uuid: Optional[str], server_uuid: Optional[str]
+) -> HttpResponse:
+    """
+    Once implemented, this function, shared between remote_billing_plans_realm
+    and remote_billing_plans_server, will return a Plans page, adjusted depending
+    on whether the /realm/... or /server/... endpoint is being used
+    """
+
+    return render_tmp_remote_billing_page(request)
+
+
+@typed_endpoint
+def remote_billing_plans_realm(request: HttpRequest, *, realm_uuid: PathOnly[str]) -> HttpResponse:
+    return remote_billing_plans_common(request, realm_uuid=realm_uuid, server_uuid=None)
+
+
+@typed_endpoint
+def remote_billing_plans_server(
+    request: HttpRequest, *, server_uuid: PathOnly[str]
+) -> HttpResponse:
+    return remote_billing_plans_common(request, server_uuid=server_uuid, realm_uuid=None)
+
+
+def remote_billing_page_common(
+    request: HttpRequest, realm_uuid: Optional[str], server_uuid: Optional[str]
+) -> HttpResponse:
+    """
+    Once implemented, this function, shared between remote_billing_page_realm
+    and remote_billing_page_server, will return a Billing page, adjusted depending
+    on whether the /realm/... or /server/... endpoint is being used
+    """
+
+    return render_tmp_remote_billing_page(request)
+
+
+@typed_endpoint
+def remote_billing_page_server(request: HttpRequest, *, server_uuid: PathOnly[str]) -> HttpResponse:
+    return remote_billing_page_common(request, server_uuid=server_uuid, realm_uuid=None)
+
+
+@typed_endpoint
+def remote_billing_page_realm(request: HttpRequest, *, realm_uuid: PathOnly[str]) -> HttpResponse:
+    return remote_billing_page_common(request, realm_uuid=realm_uuid, server_uuid=None)
```

Comment:
I think we probably want to add a decorator to all these endpoints enforcing that these URLs are all on `SELF_HOSTING_SUBDOMAIN`, a new default-None setting that we can create with perhaps a `selfhosting` value defined in `dev_settings.py`. We should just have them 404 if this isn't the right subdomain, since that is what would happen if this feature didn't exist and you tried to visit these URLs on a random Zulip server.

We might also need to add checks in places like `zerver/middleware.py` where we call into `SOCIAL_AUTH_SUBDOMAIN`, e.g. this block:
```
        subdomain = get_subdomain(request)                                                              
        if subdomain == settings.SOCIAL_AUTH_SUBDOMAIN:                                                 
            # Realms are not supposed to exist on SOCIAL_AUTH_SUBDOMAIN.                                
            return None                                                                                 
```

So maybe it makes sense for this bit to be a preparatory commit defining a `SELF_HOSTING_SUBDOMAIN` setting, adding checks like that, and then this file can define a little decorator to 404 if you're not on that domain that we'd use on all these methods.

Reply:
Added the prep commit and decorator